File: nodejs-tools/nodejs-compiler-tools/formatter-xml/src/format_xml_expression.ts
import {
  XmlCommentNode,
  XmlExpression,
  XmlPrologNode,
  XmlTagNode,
  XmlTextNode,
} from "@vighnesh153/parser-xml";
import { formatXmlPrologNode } from "./format_xml_prolog_node.ts";
import { formatXmlTagNode } from "./format_xml_tag_node.ts";
import { formatCommentNode } from "./format_comment_node.ts";
import { formatTextNode } from "./format_text_node.ts";

type FormatXmlExpressionProps = {
  readonly expression: Readonly<XmlExpression>;
  readonly indentation: number;
  readonly indentationLevel: number;
  readonly sortAttributes: boolean;
};

export function formatXmlExpression({
  expression,
  indentation,
  indentationLevel,
  sortAttributes,
}: FormatXmlExpressionProps): string {
  if (expression instanceof XmlPrologNode) {
    return formatXmlPrologNode({
      xmlPrologNode: expression,
      indentationLevel,
      indentation,
    });
  }
  if (expression instanceof XmlTagNode) {
    return formatXmlTagNode({
      xmlTagNode: expression,
      indentationLevel,
      indentation,
      sortAttributes,
    });
  }
  if (expression instanceof XmlCommentNode) {
    return formatCommentNode({
      commentNode: expression,
      indentationLevel,
      indentation,
    });
  }
  if (expression instanceof XmlTextNode) {
    return formatTextNode({
      textNode: expression,
      indentationLevel,
      indentation,
    });
  }
  throw new Error(`Unexpected xml expression: ${expression.toString(0)}`, {
    cause: expression,
  });
}

File: jvm-tools/XmlFormatter/src/main/kotlin/dev/vighnesh153/xml/formatter/fmt/FormatXmlExpression.kt
package dev.vighnesh153.xml.formatter.fmt

import dev.vighnesh153.xml.formatter.ast.XmlCommentNode
import dev.vighnesh153.xml.formatter.ast.XmlExpression
import dev.vighnesh153.xml.formatter.ast.XmlPrologNode
import dev.vighnesh153.xml.formatter.ast.XmlTagNode
import dev.vighnesh153.xml.formatter.ast.XmlTextNode

internal fun XmlExpression.format(
    indentationLevel: Int,
    indentation: Int,
    sortAttributes: Boolean,
): String = when (this) {
    is XmlPrologNode -> this.format(
        indentationLevel = indentationLevel,
        indentation = indentation
    )

    is XmlTagNode -> this.format(
        indentationLevel = indentationLevel,
        indentation = indentation,
        shouldSortAttributes = sortAttributes
    )

    is XmlCommentNode -> this.format(
        indentationLevel = indentationLevel,
        indentation = indentation
    )

    is XmlTextNode -> this.format(
        indentationLevel = indentationLevel,
        indentation = indentation
    )

    else -> throw Error("Unexpected xml expression '${toString(0)}'")
}